resources/js/components/folders.js

```diff
@@ -0,0 +1,259 @@
+                return this.checked.includes(node.id);
+            }
+
+            return (node.children || []).every(child => this.isChecked(child));
+        },
+        isIndeterminate(node) {
+            if (this.isLeaf(node)) return false; // Leaf nodes can't be indeterminate
+            const childStates = (node.children || []).map(child => this.isChecked(child) || this.isIndeterminate(child));
+            const allChecked = childStates.every(state => state);
+            const anyChecked = childStates.some(state => state);
+            return anyChecked && !allChecked;
+        },
+        updateParentStates(node) {
+            if (!node.parent) return;
+            const parent = node.parent;
+            const childIds = parent.children.map(child => child.id);
+            const selectedChildren = childIds.filter(id => this.checked.includes(id));
+
+            if (selectedChildren.length === 0) {
+                this.checked = this.checked.filter(id => id !== parent.id);
+            } else if (selectedChildren.length === childIds.length) {
+                if (!this.checked.includes(parent.id)) {
+                    this.checked.push(parent.id);
+                }
+            }
+
+            this.updateParentStates(parent);
+        },
+        toggleSelect(node) {
+            this.selected?.id === node.id ? this.unselect() : this.select(node);
+        },
+        select(node) {
+            this.$dispatch('folder-tree-select', node, this.selected);
+
+            this.selected = node;
+            if (this.selectedCallback) {
+                this.selectedCallback(node, this.selected);
+            }
+        },
+        unselect() {
+            this.selected = null;
+        },
+        removeNode(node) {
+            const nodeId = typeof node === 'object' ? node.id : node;
+
+            const traverseAndRemove = (nodes, parent = null) => {
+                for (let i = 0; i < nodes.length; i++) {
+                    if (nodes[i].id === nodeId) {
+                        nodes.splice(i, 1);
+                        return true; // Node found and removed
+                    }
+
+                    if (nodes[i][this.childrenAttribute]) {
+                        const childNodes = nodes[i][this.childrenAttribute];
+                        if (traverseAndRemove(childNodes, nodes[i])) {
+                            // Remove the parent node if it has no more children
+                            if (childNodes.length === 0) {
+                                delete nodes[i][this.childrenAttribute];
+                            }
+                            return true;
+                        }
+                    }
+                }
+                return false; // Node not found
+            };
+
+            traverseAndRemove(this.tree);
+            this.unselect();
+        },
+        addFolder(node = null, attributes = {}) {
+            let id = Math.random().toString(36).substr(2, 9);
```

Comment:
I'm not sure if this generates id that are extremely unlikely to exist in the given tree.